### PR TITLE
WIP: feat: sniper mode (implements #1284)

### PR DESCRIPTION
**Pull request by tdurieux**
- [ ] AbstractInvocation
- [ ] AnnotationFieldAccess
- [x] Annotation
- [ ] AnnotationMethod
- [ ] ArrayAccess
- [ ] Assert
- [ ] Assignment
- [ ] BinaryOperator
- [ ] BodyHolder
- [ ] Case
- [ ] Catch
- [ ] Class
- [x] Comment
- [ ] Conditional
- [ ] ConstructorCall
- [ ] Constructor
- [ ] Do
- [ ] Element
- [ ] Enum
- [ ] Executable
- [ ] ExecutableReferenceExpression
- [ ] Expression
- [ ] FieldAccess
- [ ] FieldReference
- [ ] ForEach
- [ ] For
- [ ] FormalTypeDeclarer
- [ ] If
- [ ] IntersectionTypeReference
- [ ] Invocation
- [ ] JavaDoc
- [ ] JavaDocTag
- [ ] LabelledFlowBreak
- [ ] Lambda
- [ ] Literal
- [ ] Method
- [ ] Modifiable
- [ ] MultiTypedElement
- [ ] NamedElement
- [ ] NewArray
- [ ] NewClass
- [ ] OperatorAssignment
- [ ] Package
- [ ] Parameter
- [ ] Return
- [ ] RHSReceiver
- [ ] Statement
- [ ] StatementList
- [ ] Switch
- [ ] Synchronized
- [ ] TargetedExpression
- [ ] Throw
- [ ] Try
- [ ] TryWithResource
- [ ] TypedElement
- [ ] TypeInformation
- [ ] Type
- [ ] TypeParameterReference
- [ ] TypeReference
- [ ] UnaryOperator
- [ ] VariableAccess
- [ ] Variable
- [ ] While

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
Interesting feature! And there is a lot of work in it! Thank You for that effort!
I have one idea, how to profit from that code even more... could You replace all places where you create an Action and pushes it into stack by call of appropriate `Factory.Change()#onXxxYyy(...)` method?
Example:
Replace this code
```java
 	public <C extends CtTypedElement> C setType(CtTypeReference<T> type) {
  		if (type != null) {
  			type.setParent(this);
  		}
		if (getFactory().getEnvironment().buildStackChanges()) {
			getFactory().getEnvironment().pushToStack(new UpdateAction(new ObjectContext(this, "type"), type, this.type));
		}
  		this.type = type;
  		return (C) this;
  	}
```
by
```java
 	public <C extends CtTypedElement> C setType(CtTypeReference<T> type) {
  		if (type != null) {
  			type.setParent(this);
  		}
		getFactory().Change().onObjectUpdate(this, "type", type, this.type);
  		this.type = type;
  		return (C) this;
  	}
``` 
Why that?
* The code of Element implementations is shorter and cleaner - will be easier to process it automatically in future
* The implementation of onObjectUpdate (and all similar methods which would be then needed too) might be easily changed with another implementation or extended implementation - it would be then very easy and elegant way how to implement #1168 (Solution 2 - CtElement#modelVersion attribute)

WDYT?

**Comment by pvojtechovsky** [comment]
Super! I am looking forward to having this PR in spoon! It unlocks another great use cases!
I just hope that you already convinced Martin, that this step is enough baby ;-)

> I used the sniper mode to make the change.

I am not sure what `sniper mode` means... Did I understood well that you can update origin source code, by applying only changes, without touching old code (without reformatting) ?

**Comment by tdurieux** [comment]
This PR contains a lot of bugs, we start this PR more than one year ago. 
It is not always easy to follow the change.

> I just hope that you already convinced Martin, that this step is enough baby ;-)

3000 added line is not that big ;) 

> Did I understood well that you can update origin source code, by applying only changes, without touching old code (without reformatting) ?

Exactly, it will never work  for every case but for most the case it will work (rename variable, add methods, change statement, change invocations, ...)

**Comment by pvojtechovsky** [comment]
>>  can update origin source code

> Exactly

Cool!!! 👍

File: src/main/java/spoon/reflect/visitor/printer/sniper/SniperElement.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+package spoon.reflect.visitor.printer.sniper;
+
+import spoon.diff.Action;
+import spoon.diff.AddAction;
+import spoon.diff.DeleteAction;
+import spoon.diff.DeleteAllAction;
+import spoon.diff.UpdateAction;
+
+public interface SniperElement {
+	AddAction onAdd(AddAction action);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/visitor/printer/sniper/SniperElement.java]: throw an exception if action not handled

signature return void or SniperElement

**Comment by monperrus** [comment]
still rebasable after #1377, see http://github.com/monperrus/spoon/tree/feat_sniper

**Comment by tdurieux** [comment]
I did not push the last version of this branch.

**Comment by pvojtechovsky** [comment]
I just read this code. There is really a lot of work in it and looks like it still needs some work. I personally would like this feature a lot!! It would increase usability of Spoon in real life projects, where is not acceptable to change formatting/style of whole file just because one line is added.

I suggest to extract part of this PR and merge it separately. I am speaking about all calls of new `Change()` factory like: `getFactory().Change().onXxx(...)`
The implementation of `onXxx` methods might be empty at the beginning and it would be nice if Spoon would allow client to hook at this calls easily.
I think that this part of PR is finished and there is nothing todo. And these `onXxx` calls might be used to detect changes in the model and to implement other nice features ... (improve performance, thanks to better caching, solve some features, etc....)
WDYT?

**Comment by tdurieux** [comment]
I agree with you.
I have still to fix the roles, update the contact checker and add listener support in factory.Change()

Martin is also not a big fan of onXxx. I didn't really understand what he wanted to do

**Comment by pvojtechovsky** [comment]
Before I read this PR I though about different implementation of this feature. May be it is bad idea and it cannot work, but actually it seems to me like it would be less effort to implement it.

The idea is like this
* do not collect each action on the model, but just detect if the element was modified since it's reading from sources. Let's mark modified element `ele.dirty=true` and let's mark element whose any child is dirty `ele.anyChildDirty=true`.

The pretty printer would use this algorithm:
* if the element `ele.dirty==false && ele.anyChildDirty==false', then do not scan children of this element, but simply copy all the characters from the source file from ele.position.start to ele.position.end
* if the element `ele.dirty==false && ele.anyChildDirty==true', then scan children until first dirty child is found. Then copy all the characters from the source file from ele.position.start to firstDirtyEle.position.start, then "somehow" ;-) render dirty element and search for next dirty child element ... copy not modified sources and render modified ...
* the removed elements would be removed from printed result, because these elements are not found by scanning and therefore it's source is not copied from old to new content.
* the added elements would be pretty printed using existing printer.
* the moved elements might move too, because their source is copied on the new position

If that algorithm would work, then it would be much less work, because we do not have to care about each possible operation on the model.
WDYT?

**Comment by tdurieux** [comment]
The first problem that I found is:
If you change the modifier or the name of a class you will reprint the complete class. 

And You will still have the issue to know where to print the new element (no position).


 The goal of this approach is to minify the resulting diff

**Comment by pvojtechovsky** [comment]
> Martin is also not a big fan of onXxx.
Optionally I would like one method `Factory#onChange(operationType, role, target, oldValue, newValue, ...?...)`.

> If you change the modifier or the name of a class you will reprint the complete class.
yes, it is the bigger challenge of this approach - to detect headers, footers and separators of the elements which contains other elements and in this case to update only class header.

**Comment by monperrus** [comment]
> Interesting feature! And there is a lot of work in it

Indeed. The will probably a long-standing WIP PR, probably to be merged this Fall. The goal of having it here is to keep it constantly on sync with "master", while progressing on it.

> Martin is also not a big fan of onXxx. 

There is a misunderstanding, I think it's a promising direction.

> I suggest to extract part of this PR and merge it separately. I am speaking about all calls of new `Change()` factory like: `getFactory().Change().onXxx(...)`

Can we use the sniper mode itself to do this from to extract such a new PR from this PR? :-)

**Comment by tdurieux** [comment]
@leventov I changed to `ChangeFactory.java` what do you think?

**Comment by pvojtechovsky** [comment]
ChangeFactory.java is ok for me

**Comment by pvojtechovsky** [comment]
Another idea how to implement pretty printer, which produces sources with minimum differences to origin sources... To collect and remember separators between elements in origin sources and then use them during pretty printing.

* no need to detect changes at all
* the position of each element would be extended with information about prefix and suffix separator
* the current pretty printer would use these separators

I am thinking about optional ways how to do it, because I have feeling like current approach, which applies all the model changes separatelly is much more complicated and may be for some combinations of actions impossible.

Side effect: If we would have separators stored in model, then clients might influence the formatting of generated code (e.g code generated from template would keep formatting of the template) ... I guess clients would like it too.

**Comment by placeybordeaux** [comment]
This branch appears to be ~400 commits behind at this point. As a new comer to the project I couldn't figure out how to merge the changes cleanly, but something like sniper support, or better support for outputting code that conforms to some style guide is a must have to use for my purposes.

Is this still the best chance for 'sniper' support?

**Comment by pvojtechovsky** [comment]
@placeybordeaux The PR #1927 tries to implement the same feature using different concept. Try it

**Comment by monperrus** [comment]
Closed per #1927. Thanks Thomas for all your inputs and ideas on this topic.

**Pull request closed without merging**